airlock/management/commands/run_file_uploader.py

```diff
@@ -0,0 +1,153 @@
+from opentelemetry import trace
+
+import old_api
+from airlock.business_logic import bll
+from airlock.enums import RequestStatus
+from airlock.users import User
+from services.tracing import instrument
+
+
+logger = logging.getLogger(__name__)
+
+# We need a user with the output-checker role to access some
+# bll methods
+system_user = User("system", output_checker=True)
+
+
+class Command(BaseCommand):
+    """
+    Poll database for approved requests that still have files pending upload
+    """
+
+    def add_arguments(self, parser):
+        # In production, we want this check to run forever. Using a
+        # function means that we can test it on a finite number of loops.
+        parser.add_argument("--run-fn", default=lambda: True, help=argparse.SUPPRESS)
+
+    def handle(self, *args, **options):
+        run_fn = options["run_fn"]
+
+        logger.debug("Watching for tasks")
+
+        while run_fn():  # pragma: no branch
+            # Find approved requests
+            approved_requests = bll.get_approved_requests(user=system_user)
+
+            if not approved_requests:
+                # No pending file uploads found; wait for UPLOAD_DELAY seconds
+                # before checking again
+                time.sleep(settings.UPLOAD_DELAY)
+                continue
+
+            for approved_request in approved_requests:
+                # Find incomplete file uploads for this request
+                # This retrieves ALL incomplete file uploads, including those that have
+                # been reached the max attempt limit, so we can check for any
+                # requests that should be set to released
+                files_for_upload = get_upload_files_and_update_request_status(
+                    approved_request
+                )
+                if not files_for_upload:
+                    # All files are now uploaded, the status has been updated to released,
+                    # nothing to do
+                    continue
+
+                workspace = bll.get_workspace(approved_request.workspace, system_user)
+                for file_for_upload in files_for_upload:
+                    if file_for_upload.upload_attempts >= settings.UPLOAD_MAX_ATTEMPTS:
+                        logger.debug(
+                            "Max upload attempts reached for %s - %s, skipping",
+                            approved_request.id,
+                            file_for_upload.relpath,
+                        )
+                        continue
+                    # increment the retry attempts; if something goes wrong, we still
+                    # want this to be updated
+                    file_for_upload = bll.register_file_upload_attempt(
+                        approved_request, file_for_upload.relpath
+                    )
+
+                    tracer = trace.get_tracer(
+                        os.environ.get("OTEL_SERVICE_NAME", "airlock")
+                    )
```

Comment:
I'm quite otel-naive so this may be a silly question, but should this get instantiated higher up, either at the top of the file or the top of the `handle()` method? Maybe it's a super cheap operation, but it just looks a bit weird to me that we're creating a new tracer object from an environment variable for each individual file.

Reply:
yes, I think you're right.  I took this from the `@instrument` decorator, which I couldn't use just as it was, because I needed to add the exception event to it, and that instantiates the tracer in the decorator. But I think we should do this before the while loop, will update

Reply:
Ah yeah, but looks like the decorator instantiates it at function definition time so it's still only happening once.

Reply:
Yes, I know, but I missed that when I added it in the first place